### PR TITLE
Made get_conn in JdbcHook threadsafe to avoid OSError: JVM is already started

### PR DESCRIPTION
**Pull request by dabla**
When using our in house [IterableOperator](https://github.com/apache/airflow/pull/42572/) to allow easy multithreading with operators within the same worker, we discovered a thread safety issue with the JdbcHook.  As multiple threads where trying to get a connection through the jaydebeapi libary, we sometimes got random "OSError: JVM is already started" errors, due to the high concurrency/parallelism, something you won't experience if you do the same using the expand mapped task as each task instance is then run on a different Python process.

To fix the issue I made the get_conn method of the JdbcHook synchronized and added a unit test which simulates the behaviour to avoid regression in the future so this is well documented through a test.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: providers/tests/jdbc/hooks/test_jdbc.py [hunk 1]
```diff
@@ -54,10 +56,17 @@ def get_hook(
             **conn_params,
         }
     )
+    jvm_started = False
 
     class MockedJdbcHook(JdbcHook):
         @classmethod
         def get_connection(cls, conn_id: str) -> Connection:
+            # nonlocal jvm_started
```
Comment (potiuk) [inline providers/tests/jdbc/hooks/test_jdbc.py]: Is it commented intentionally ?

Comment (dabla) [reply providers/tests/jdbc/hooks/test_jdbc.py]: Yeah forgot to remove it apologies

File: providers/src/airflow/providers/jdbc/hooks/jdbc.py [hunk 2]
```diff
@@ -25,6 +25,7 @@
 import jaydebeapi
 import jpype
 from sqlalchemy.engine import URL
+from wrapt import synchronized
```
Comment (potiuk) [inline providers/src/airflow/providers/jdbc/hooks/jdbc.py]: I have not found "synchronized" as something that wrapt would publish as an interface. It seems it is somewhat accidentally available in wrapt implementation and might change any time. 

I think we should find a more "official" tool to do it.

Comment (potiuk) [reply providers/src/airflow/providers/jdbc/hooks/jdbc.py]: Hmm. I found it in https://wrapt.readthedocs.io/en/master/examples.html - but search did not found it. It's therefore an example in wrapt - which is also a bit strange to rely on ?

Comment (potiuk) [reply providers/src/airflow/providers/jdbc/hooks/jdbc.py]: Possibly better to use with 

```
self.lock = threading.RLock()

with self.lock:
     jaydebeapi.connect( ...
```

It's essentially the same but better expresses the fact that we want to synchronize jaydebeapi, not the `get_conn` method.

Comment (dabla) [reply providers/src/airflow/providers/jdbc/hooks/jdbc.py]: > Hmm. I found it in https://wrapt.readthedocs.io/en/master/examples.html - but search did not found it. It's therefore an example in wrapt - which is also a bit strange to rely on ?

I liked the solution as to me, coming from Java, it seemed like a nice and clean solution.  You can use the decorator as an aspect which prevents other threads from entering until the current one has finished, no need to hard code this into the function as this is a crosscutting concern, imho.

Comment (dabla) [reply providers/src/airflow/providers/jdbc/hooks/jdbc.py]: > Possibly better to use with
> 
> ```
> self.lock = threading.RLock()
> 
> with self.lock:
>      jaydebeapi.connect( ...
> ```
> 
> It's essentially the same but better expresses the fact that we want to synchronize jaydebeapi, not the `get_conn` method.

I will adapt it like this then.

**Review by potiuk — APPROVED** [review]

File: providers/src/airflow/providers/jdbc/hooks/jdbc.py (hunk 2 above)
Comment (potiuk) [reply providers/src/airflow/providers/jdbc/hooks/jdbc.py]: > I liked the solution as to me, coming from Java, it seemed like a nice and clean solution.

Yeah. It's nice, but it's more of a byproduct of what they have as an example usage. What happens if one day they release the package and the "synchronized" decorator will not be there or moved? 

They can do it even in patchlevel version. and claim "it's ok to move it - it's just an example" - we cannot rely on it being there. 

We could also write our own decorator (basically copy what they have done) and keep it in, but that seems an overkill if we can do the same with `self.lock =" and `with lock` two lines added. 

It's abit like infamous `leftpad` fiasco.

**Pull request merged**